core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -773,16 +773,26 @@ class DAGScheduler(
     stage.pendingTasks.clear()
 
     // First figure out the indexes of partition ids to compute.
-    val partitionsToCompute: Seq[Int] = {
+    val (allPartitions: Seq[Int], partitionsToCompute: Seq[Int]) = {
       stage match {
         case stage: ShuffleMapStage =>
-          (0 until stage.numPartitions).filter(id => stage.outputLocs(id).isEmpty)
+          val allPartitions = 0 until stage.numPartitions
+          val filteredPartitions = allPartitions.filter(id => stage.outputLocs(id).isEmpty)
+          (allPartitions, filteredPartitions)
         case stage: ResultStage =>
           val job = stage.resultOfJob.get
-          (0 until job.numPartitions).filter(id => !job.finished(id))
+          val allPartitions = 0 until job.numPartitions
+          val filteredPartitions = allPartitions.filter(id => !job.finished(id))
+          (allPartitions, filteredPartitions)
       }
     }
 
+    // Reset internal accumulators only if this stage is not partially submitted
+    // Otherwise, we may override existing accumulator values from some tasks
```

Comment:
An accumulator value counts only if the associated task was successful. This means as long as a single task in the stage was successful, then we can't re-initialize the set of internal accumulators because doing so will wipe out the partial values from the succeeded tasks. This corresponds to (3) and (4) in the cases you listed.

The relevant test is in `AccumulatorSuite`.